zkevm-circuits/src/compress_circuit.rs

```diff
@@ -0,0 +1,309 @@
+//! The compress circuit implementation.
+
+use crate::evm_circuit::util::from_bytes;
+use crate::util::{Expr, SubCircuitConfig};
+use bit_vec::BitVec;
+use eth_types::Field;
+use halo2_proofs::{
+    circuit::{AssignedCell, Layouter, Region, Value},
+    plonk::{Advice, Column, ConstraintSystem, Error, Selector, TableColumn},
+    poly::Rotation,
+};
+use lazy_static::lazy_static;
+use std::marker::PhantomData;
+
+lazy_static! {
+    // TODO: maybe fixed huffman table will get from witness
+    static ref FIXED_HUFFMAN_TABLE: [BitVec; 256] = [(); 256].map(|_| BitVec::new());
+}
+
+mod combine {
+    const LEFT_SHIFT_8BITS: u64 = 256;
+
+    use crate::util::Expr;
+    use eth_types::Field;
+    use halo2_proofs::plonk::Expression;
+    pub(crate) fn expr<F: Field>(parts: [Expression<F>; 32]) -> Expression<F> {
+        parts
+            .into_iter()
+            .fold(0.expr(), |acc, part| acc * F::from(LEFT_SHIFT_8BITS) + part)
+    }
```

Comment:
The field we're working can only store ~253 bits, so this will overflow which makes it not safe to use. To compare 32byte values you could instead rlc encode them instead.

Reply:
Because the 32byte is split from raw_public_inputs's row which does not overflow, so I think just combining the 32byte immediately is OK?

Anyway, I agree that using rlc encoding will be better

Reply:
I don't think so unless you add additional constraints. If the prover can input a full byte value for each part it's always possible to overflow. If some bytes are checked in the smart then you would have to check byte per byte instead of the sum of the bytes if there can be overflow there.